src/crystal/system/unix/evented/event_loop.cr

```diff
@@ -0,0 +1,506 @@
+  # big without a good reason (i.e. we need that many fds at that time). This
+  # assumption allows the arena to not have to keep a list of free indexes.
+  protected class_getter arena = Arena(PollDescriptor).new(max_fds)
+
+  private def self.max_fds : Int32
+    if LibC.getrlimit(LibC::RLIMIT_NOFILE, out rlimit) == -1
+      raise RuntimeError.from_errno("getrlimit(RLIMIT_NOFILE)")
+    end
+    rlimit.rlim_max.clamp(..Int32::MAX).to_i32!
+  end
+end
+
+# Polling EventLoop.
+#
+# This is the abstract interface that implements `Crystal::EventLoop` for
+# polling based UNIX targets, such as epoll (linux), kqueue (bsd), or poll
+# (posix) syscalls. This class only implements the generic parts for the
+# external world to interact with the loop. A specific implementation is
+# required to handle the actual syscalls.
+#
+# The event loop registers the fd into the kernel data structures when an IO
+# operation would block, then keeps it there until the fd is closed.
+#
+# NOTE: the fds must have `O_NONBLOCK` set.
+#
+# It is possible to have multiple event loop instances, but an fd can only be in
+# one instance at a time. When trying to block from another loop, the fd will be
+# removed from its associated loop and added to the current one (this is
+# automatic). Trying to move a fd to another loop with pending waiters is
+# unsupported and will raise an exception. See `PollDescriptor#release`.
+#
+# A timed event such as sleep or select timeout follows the following logic:
+#
+# 1. create an `Event` (actually reuses it, see `FiberChannel`);
+# 2. register the event in `@timers`;
+# 3. supend the current fiber.
+#
+# The timer will eventually trigger and resume the fiber.
+# When an IO operation on fd would block, the loop follows the following logic:
+#
+# 1. register the fd (once);
+# 2. create an `Event`;
+# 3. suspend the current fiber;
+#
+# When the IO operation is ready, the fiber will eventually be resumed (one
+# fiber at a time). If it's an IO operation, the operation is tried again which
+# may block again, until the operation succeeds or an error occured (e.g.
+# closed, broken pipe).
+#
+# If the IO operation has a timeout, the event is also registered into `@timers`
+# before suspending the fiber, then after resume it will raise
+# `IO::TimeoutError` if the event timed out, and continue otherwise.
+#
+# OPTIMIZE: collect fibers & canceled timers, delete canceled timers when
+# processing timers, and eventually enqueue all fibers; it would avoid repeated
+# lock/unlock timers on each #resume_io and allow to replace individual fiber
+# enqueues with a single batch enqueue (simpler).
+abstract class Crystal::Evented::EventLoop < Crystal::EventLoop
+  {% if flag?(:preview_mt) %}
+    @run_lock = Atomic::Flag.new # protects parallel runs
+  {% end %}
+
+  def initialize
+    @lock = SpinLock.new # protects parallel accesses to @timers
+    @timers = Timers.new
+  end
```

Comment:
**suggestion:** Move these ivar initializers outside of `#initialize`. Then implementing classes don't need to call `super`.